pkg/ucp/store/postgres/postgresclient.go

```diff
@@ -0,0 +1,422 @@
+	if errors.Is(err, pgx.ErrNoRows) {
+		return nil, &store.ErrNotFound{ID: id}
+	} else if err != nil {
+		return nil, err
+	}
+
+	return &obj, nil
+}
+
+// Query implements store.StorageClient.
+func (p *PostgresClient) Query(ctx context.Context, query store.Query, options ...store.QueryOptions) (*store.ObjectQueryResult, error) {
+	if ctx == nil {
+		return nil, &store.ErrInvalid{Message: "invalid argument. 'ctx' is required"}
+	}
+
+	err := query.Validate()
+	if err != nil {
+		return nil, &store.ErrInvalid{Message: fmt.Sprintf("invalid argument. Query is invalid: %s", err.Error())}
+	}
+
+	config := store.NewQueryConfig(options...)
+
+	// For a scope query, we need to perform the same normalization as we do for other operations on scopes.
+	resourceType := storeutil.NormalizePart(query.ResourceType)
+	if query.IsScopeQuery {
+		var err error
+		resourceType, err = storeutil.ConvertScopeTypeToResourceType(query.ResourceType)
+		if err != nil {
+			return nil, err
+		}
+
+		resourceType = storeutil.NormalizePart(resourceType)
+	}
+
+	var routingScopePrefixFilter *string
+	if query.RoutingScopePrefix != "" {
+		routingScopePrefixFilter = to.Ptr(storeutil.NormalizePart(query.RoutingScopePrefix))
+	}
+
+	var timestampFilter *string
+	if config.PaginationToken != "" {
+		ts, err := p.parsePaginationToken(config.PaginationToken)
+		if err != nil {
+			return nil, &store.ErrInvalid{Message: "invalid argument. 'query.PaginationToken' is invalid."}
+		}
+		timestampFilter = &ts
+	}
+
+	var limitFilter *int
+	if config.MaxQueryItemCount > 0 {
+		limitFilter = &config.MaxQueryItemCount
+	}
+
+	// For a scope query, we need to perform the same normalization as we do for other operations on scopes.
+	if query.IsScopeQuery {
+		var err error
+		query.ResourceType, err = storeutil.ConvertScopeTypeToResourceType(query.ResourceType)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// NOTE: building SQL by concatenating strings is hard to do safely and should be avoided.
+	// If you need to work on this code MAKE SURE you use SQL parameters
+	// for any user input.
+	sql := `
+SELECT original_id, etag, resource_data, created_at 
+FROM resources
+WHERE ((root_scope = $1) OR ($2 AND (root_scope LIKE $1 || '%'))) AND 
+	resource_type = $3 AND 
+	((routing_scope LIKE $4 || '%') OR $4 IS NULL) AND 
+	(created_at > $5::TIMESTAMP OR $5 IS NULL)
+ORDER BY created_at ASC
+LIMIT $6`
+
+	args := []any{
+		// If ScopeRecursive is false, the RootScope must match exactly.
+		// If ScopeRecursive is true, the RootScope must be a prefix of the stored RootScope.
+		storeutil.NormalizePart(query.RootScope),
+		query.ScopeRecursive,
+		resourceType,
+		routingScopePrefixFilter, // RoutingScopePrefix is optional and always treated as as prefix.
+		timestampFilter,          // Optional for pagination.
+		limitFilter,              // NOTE: Postgres allows LIMIT to be set with a NULL value to mean no limit.
```

Comment:
Should the search be case insensitive?

Reply:
Search of what? Resource id are case-insensitive. That's how they work. The code has numerous places where the behavior is explained. Please let me know if you have suggestions about how this could be improved.